http/server/logging/logging.go

```diff
@@ -4,22 +4,34 @@ import (
 	"context"
 	"log/slog"
 	"net/http"
+	"reflect"
 	"time"
 
 	"github.com/gorilla/mux"
 	"google.golang.org/grpc/metadata"
 )
 
+type initFunc func(w http.ResponseWriter, r *http.Request) map[string]interface{}
+type loggingFunc func(w http.ResponseWriter, r *http.Request, attrs map[string]interface{}) map[string]interface{}
+type CustomAttributes struct {
+	AttributeInitializer *initFunc    // sets keys for custom attributes at the beginning of ServeHTTP()
+	AttributeAssigner    *loggingFunc // assigns values for custom attributes after request has completed
+}
+
 // TODO (Tom): Add a logger wrapper in its own package
 // https://medium.com/@ansujain/building-a-logger-wrapper-in-go-with-support-for-multiple-logging-libraries-48092b826bee
-
 // more info about http handler here: https://pkg.go.dev/net/http#Handler
-func NewLogging(logger *slog.Logger) mux.MiddlewareFunc {
+
+// If source is empty, it will be set to "ApiRequestLog"
+// If ANY fields in customAttributeAssigner are empty, or the struct itself is empty, extra attributes will not be assigned
+func NewLogging(logger *slog.Logger, source string, customAttributeAssigner CustomAttributes) mux.MiddlewareFunc {
 	return func(next http.Handler) http.Handler {
 		return &loggingMiddleware{
-			next:   next,
-			now:    time.Now,
-			logger: *logger,
+			next:                    next,
+			now:                     time.Now,
+			logger:                  *logger,
+			source:                  source,
+			customAttributeAssigner: customAttributeAssigner,
```

Comment:
Do you support the initializer and assigner funcs to be nil? If not then you should assign them to default functions if not provided